docs/design/trimmed-assemblies.md

```diff
@@ -0,0 +1,212 @@
+
+### Examples
+
+This shows how a developer could turn on aggressive trimming for framework assemblies (which are defined to be `IsTrimmable` by the SDK):
+
+```xml
+<PropertyGroup>
+  <TrimMode>link</TrimMode>
+</PropertyGroup>
+```
+
+This shows how Blazor (or a developer) could hook into the build to opt assemblies into different levels of trimming based on the filename:
+
+```xml
+<Target Name="PrepareForBlazorILLink"
+        BeforeTargets="PrepareForILLink">
+  <PropertyGroup>
+    <!-- Set the default TrimMode for IsTrimmable assemblies -->
+    <TrimMode>link</TrimMode>
+  </PropertyGroup>
+  <ItemGroup>
+    <ManagedAssemblyToLink Condition="'$([System.String]::Copy('%(ManagedAssemblyToLink.Filename)').StartsWith('Microsoft.AspNetCore.'))">
+      <!-- Trim these assemblies using the global TrimMode -->
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+    <ManagedAssemblyToLink Condition="'$([System.String]::Copy('%(ManagedAssemblyToLink.Filename)').StartsWith('ThirdPartyAssembly.'))">
+      <!-- Trim these assemblies with assembly-level trimming. Implies IsTrimmable. -->
+      <TrimMode>copyused</TrimMode>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+### Other options
+
+.NET 5 introduced a host of additional SDK options that map directly to the underlying illink options. The full list is documented at https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options.
+
+## .NET 6
+
+### `AssemblyMetadata("IsTrimmable", "True")`
+
+An assembly-level `AssemblyMetadataAttribute` may be placed on an assembly to indicate that it can be trimmed:
+
+```csharp
+[assembly: AssemblyMetadata("IsTrimmable", "True")]
+```
+
+The behavior is the same as the `IsTrimmable` MSBuild metadata, so that:
+- Assemblies with this attribute are trimmed with the global `TrimMode`
+- Assemblies without this attribute are rooted, and given the `copy` action
+
+The only understood value is `True` (case-insensitive). Adding `[assembly: AssemblyMetadata("IsTrimmable", "False")]` will have no effect on the linker's behavior, because unattributed assemblies are assumed not to be trimmable by default. We will issue a warning in this case, to discourage misleading use of the attribute.
+
+The attribute will survive trimming like other assembly-level attributes do.
+
+If `IsTrimmable` MSBuild metadata is set for an assembly, this overrides the `IsTrimmable` attribute. This allows a developer to opt an assembly into trimming even if it does not have the attribute, or to disable trimming of an assembly that has the attribute.
+
+Instead of using `IsTrimmable` metadata in the SDK to control trimmable assemblies, we will move to a model where all trimmable SDK assemblies are built with `[assembly: AssemblyMetadata("IsTrimmable", "True")]`.
+
+### `TrimmableAssembly`
```

Comment:
Who do we expect to use this option? Is it necessary to build it now given that we have the `PrepareForILLink` target? It is simply a convenience.

I wonder if this should be moved into `Future evolution` and not proposed for .NET 6.

Reply:
It's purely a convenience on top of the MSBuild `IsTrimmable` metadata. I added this in response to https://github.com/mono/linker/issues/1269#issuecomment-731449888:
> I think it's crucial during the transition period to allow developers to easily experiment with enabling trimming for any assembly. It should be as simple as passing/adding msbuild property.

I'll let @marek-safar comment on whether it belongs in .NET6 or future.

Reply:
We already got reports from .net5 developers asking for an easy way to trim more assemblies. Similarly, we will most likely need it for our own usage (trimming testing frameworks) in dotnet/runtime and messing up with msbuild metadata for this is well, messy. I think we should make this operation convenient in .net6.

Reply:
Ok, as long as we have user scenarios for it, that’s fine with me.